dpcpp/base/complex.hpp

```diff
@@ -0,0 +1,275 @@
+    {
+        real_ += real;
+        return *this;
+    }
+
+    complex& operator-=(const value_type& real)
+    {
+        real_ -= real;
+        return *this;
+    }
+
+    complex& operator*=(const value_type& real)
+    {
+        real_ *= real;
+        imag_ *= real;
+        return *this;
+    }
+
+    complex& operator/=(const value_type& real)
+    {
+        real_ /= real;
+        imag_ /= real;
+        return *this;
+    }
+
+    template <typename T>
+    complex& operator+=(const std::complex<T>& val)
+    {
+        real_ += val.real();
+        imag_ += val.imag();
+        return *this;
+    }
+
+    template <typename T>
+    complex& operator-=(const std::complex<T>& val)
+    {
+        real_ -= val.real();
+        imag_ -= val.imag();
+        return *this;
+    }
+
+    template <typename T>
+    inline complex& operator*=(const std::complex<T>& val);
+
+    template <typename T>
+    inline complex& operator/=(const std::complex<T>& val);
+
+    complex& operator+=(const gko::complex<value_type>& val)
+    {
+        real_ += val.real();
+        imag_ += val.imag();
+        return *this;
+    }
+
+    complex& operator-=(const gko::complex<value_type>& val)
+    {
+        real_ -= val.real();
+        imag_ -= val.imag();
+        return *this;
+    }
+
+    inline complex& operator*=(const gko::complex<value_type>& val);
+
+    inline complex& operator/=(const gko::complex<value_type>& val);
+
+// It's for MacOS.
+// TODO: check whether mac compiler always use complex version even when real
+// half
+#define COMPLEX_HALF_OPERATOR(_op, _opeq)                                  \
+    friend complex<sycl::half> operator _op(const complex<sycl::half> lhf, \
+                                            const complex<sycl::half> rhf) \
+    {                                                                      \
+        auto a = lhf;                                                      \
+        a _opeq rhf;                                                       \
+        return a;                                                          \
+    }
+
+    COMPLEX_HALF_OPERATOR(+, +=)
+    COMPLEX_HALF_OPERATOR(-, -=)
+    COMPLEX_HALF_OPERATOR(*, *=)
+    COMPLEX_HALF_OPERATOR(/, /=)
+
+#undef COMPLEX_HALF_OPERATOR
+
+
+    complex operator-() const { return complex(-real_, -imag_); }
+
+private:
+    value_type real_;
+    value_type imag_;
+};
+
+}  // namespace gko
+
+
+// after providing std::complex<sycl::half>, we can load their <complex> to
+// complete the header chain.
+
+#if GINKGO_DPCPP_MAJOR_VERSION > 7 || \
+    (GINKGO_DPCPP_MAJOR_VERSION == 7 && GINKGO_DPCPP_MINOR_VERSION >= 1)
+
+#if defined(__has_include_next)
+// GCC/clang support go through this path.
+#include_next <complex>
+#else
+// MSVC doesn't support "#include_next", so we take the same workaround in
+// stl_wrappers/complex.
+#include <../stl_wrappers/complex>
+#endif
+
+#else
+
+
+#include <complex>
+
+
+#endif
+
```

Comment:
You should be able to just remove this and always include `<complex>`.
Some of the comments can also be adjusted.

Reply:
Yes, it will be a normal header again after everything works well